README.md

```diff
@@ -6,5 +6,73 @@
+those containers.
+
+The `cc-agent` execution unit is the pod. A `cc-agent` pod is a container sandbox
+defined by a set of namespaces (NS, UTS, IPC and PID). `cc-runtime` can run several
+containers per pod to support container engines that require multiple containers
+running inside a single VM. In the case of docker, `cc-runtime` creates a single
+container per pod.
+
+`cc-agent` uses a communication protocol defined by the [hyperstart](https://github.com/hyperhq/hyperstart)
+project. This was chosen to maintain backward compatibility with the `hyperstart`
+agent used in the 2.x `Clear Containers` architecture.
+
+The `cc-agent` interface consists of:
+- A control serial channel over which the `cc-agent` sends and receives specific
+  commands for controlling and managing pods and containers. Detailed information
+  about the commands can be found at [`cc-agent` API](https://github.com/clearcontainers/agent/tree/master/api).
+- An I/O serial channel for passing the container processes output streams (`stdout`,
+  `stderr`) back to `cc-proxy` and receiving the input stream (`stdin`) for them. As
+  all streams for all containers are going through one single serial channel, the
+  `cc-agent` prepends them with container specific sequence numbers. There are
+  at most two sequence numbers per container process: one for `stdout` and `stdin`,
+  and another one for `stderr`.
+
+`cc-agent` supports the following commands:
+- `StartPodCmd`: Sets up a pod in a newly created VM. 
+- `NewContainerCmd`: Creates a new container within a pod. This needs to be sent
+  after the `StartPodCmd` has been issued for starting a pod. This command also
+  starts the container process.
+- `ExecCmd`: Executes a new process within an already running container.
+- `KillContainerCmd`: `cc-shim` uses this to send signals to a container process.
+- `WinsizeCmd`: `cc-shim` uses this to change the terminal size of the terminal
+  associated with a container.
+- `RemoveContainerCmd`: Removes a container from the pod. This command will fail
+  if the container is in a running state.
+- `Destroypod`: Removes all containers within a pod . All containers need to be
+  in a stopped state for this command to succeed. The command also frees resources
+  associated with the pod.
+
+Each control message is composed of a command code and a payload required for
+the command:
+
+```
+  ┌────────────────┬────────────────┬──────────────────────────────┐
+  │  Command Code  │     Length     │ Payload(request or response) │
+  │   (32 bits)    │    (32 bits)   │     (data length bytes)      │
+  └────────────────┴────────────────┴──────────────────────────────┘
+```
+- `Command Code` is the predefined id associated with a command.
+- `Length` is the size of the entire message in bytes and encoded in network order.
+- `Payload` is the JSON-encoded command request or response data.
+
+Each stream message is composed of a stream sequence code and a payload containing
```

Comment:
Each stream message consists of a stream sequence code and a payload that contains